sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala

```diff
@@ -16,254 +16,459 @@
-    new Iterator[ColumnarBatch] {
-      var resultBatch: Option[ColumnarBatch] = None
+  def isEmpty: Boolean = pending.isEmpty
 
-      TaskContext.get().addTaskCompletionListener[Unit](_ => closeBatch())
+  override def close(): Unit = pending.forEach(_.close())
+}
 
-      private def closeBatch(): Unit = resultBatch.foreach(_.close())
+case class GpuOutOfCoreSortIterator(
+    iter: Iterator[ColumnarBatch],
+    sorter: GpuSorter,
+    cpuOrd: LazilyGeneratedOrdering,
+    targetSize: Long,
+    totalTime: GpuMetric,
+    sortTime: GpuMetric,
+    outputBatches: GpuMetric,
+    outputRows: GpuMetric,
+    peakDevMemory: GpuMetric,
+    spillCallback: (String, Long) => Unit) extends Iterator[ColumnarBatch]
+    with Arm with AutoCloseable {
+
+  // There are so many places where we might hit a new peak that it gets kind of complex
+  // so we are picking a few places that are likely to increase the amount of memory used.
+  private var peakMemory = 0L
+
+  private val pending = new Pending(cpuOrd)
+
+  private val sorted = new LinkedList[SpillableColumnarBatch]()
+  private var sortedSize = 0L
+
+  override def hasNext: Boolean = !sorted.isEmpty || !pending.isEmpty || iter.hasNext
+
+  // Use types for the UnsafeProjection otherwise we need to have CPU BoundAttributeReferences
+  private lazy val unsafeProjection = UnsafeProjection.create(sorter.projectedBatchTypes)
+  private lazy val converters = new GpuRowToColumnConverter(
+    TrampolineUtil.fromAttributes(sorter.projectedBatchSchema))
+
+  private def convertBoundaries(tab: Table): Array[UnsafeRow] = {
+    import scala.collection.JavaConverters._
+    val cb = withResource(new NvtxRange("COPY BOUNDARIES", NvtxColor.PURPLE)) { _ =>
+      new ColumnarBatch(
+        GpuColumnVector.extractColumns(tab, sorter.projectedBatchTypes).map(_.copyToHost()),
+        tab.getRowCount.toInt)
+    }
+    withResource(cb) { cb =>
+      withResource(new NvtxRange("TO UNSAFE ROW", NvtxColor.RED)) { _ =>
+        cb.rowIterator().asScala.map(unsafeProjection).map(_.copy().asInstanceOf[UnsafeRow]).toArray
+      }
+    }
+  }
 
-      private def loadNextBatch(): Option[ColumnarBatch] = {
-        if (batchIter.hasNext) {
-          if (singleBatchOnly && haveSortedBatch) {
-            throw new UnsupportedOperationException("Expected single batch to sort")
+  private final def splitAfterSortAndSave(sortedTbl: Table, sortedOffset: Int = -1): Unit = {
+    var memUsed: Long = GpuColumnVector.getTotalDeviceMemoryUsed(sortedTbl)
+    // We need to figure out how to split up the data into reasonable batches. We could try and do
+    // something really complicated and figure out how much data get per batch, but in practice
+    // we really only expect to see one or two batches worth of data come in, so lets optimize
+    // for that case and set the targetBatchSize to always be 1/8th the targetSize.
+    val targetBatchSize = targetSize / 8
```

Comment:
In the poorly configured case, `targetSize` could be 16KB. So this means we'd do 2KB batches here?

Reply:
Yes, but the only time that would really happen is unit tests.  We eventually might want to prevent people from shooting themselves in the foot, but for some reason I am feeling optimistic about human nature today, and I don't want to have to figure out how to make the tests that set it to 0 still work if we disallow setting it to anything smaller than say a few hundred MB.

Reply:
> for some reason I am feeling optimistic about human nature today

:)